src/07-registers/memory-mapping.md

```diff
@@ -0,0 +1,115 @@
+
+Let's assume that we have written a program that adds two numbers together. The
+compiler looks at our code and generates instructions for the processor to
+follow. The instructions generated might look like the following:
+
+* Copy the first number into register `r1`.
+* Copy the second number into register `r2`.
+* Perform the add operation using `r1` and `r2`, storing the results in register `r3`.
+
+Of course, we need somewhere to store all of our instructions and values that
+we want the instructions to act on. There aren't enough registers to store all
+of that. This is why we have **memory**.
+
+## Memory
+
+All microcontrollers are a bit different. One of those differences is the size and type
+of memory offered. Our microcontroller has two different kinds of memory: [Flash] and [SRAM].
+These sit on the microcontroller, but outside of the processor.
+
+[Flash]: https://en.wikipedia.org/wiki/Flash_memory
+[SRAM]: https://en.wikipedia.org/wiki/Static_random-access_memory
+
+<div style="width: 100%">
+<img style="display: block; margin: auto" alt="Diagram of Microcontroller with Memory" src="../assets/mmio-memory.png"></img>
+</div>
+
+Let's expand our "add two numbers" example from before, now introducing memory. Here are
+instruction that a compiler might generate for the processor to execute:
+
+1. Read from Flash address `0x00`. Store the value in processor register `r1`.
+1. Read from Flash address `0x04`. Store the value in processor register `r2`.
+1. Perform the add operation. Store the results in processor register `r3`.
+1. Write the value stored in `r3` to SRAM address `0x00`.
+
+> Each register is 32 bits in size. Each address points to a single byte (8-bits) of
+> memory. Therefore, if you ask to perform an operation to read 32-bits from memory,
+> you'll read from more than just the address you initially point to. For example,
+> reading 32-bits from address `0x00` will cause us to read not only the 8-bits in
+> `0x00`, but additionally the eight bits from `0x01`, `0x02` and `0x03`. This is why
+> our second value would be found following address `0x04`.
+
+Our processor generally provides two potential instructions for memory
+operations: one to read from an address in memory, and another to write to an
+address in memory. But if all we can provide to these instructions is an
+address, how does the processor know _which_ type of memory to use? For
+example, if we ask the processor to load from memory address `0x00`, how does
+the processor know whether to look at the first byte of Flash or the first byte
+of SRAM?
+
+The solution used by the designers of our processor is to assign all of the
+addresses from both memory types onto one global address space. For example, we
+could create a simple mapping of values that looks like this:
+
+```
+Addresses 0x0000_0000 - 0x0003_FFFF: 256kb Flash
+Addresses 0x0004_0000 - 0x0004_BFFF: 48kb SRAM
+```
+
+> This isn't the actual map for your controller, but useful as an example.
```

Comment:
Could we just use the real memory map from the F3? I don't think it's any more complicated, you can even use 0x0 for flash since it's accessible from address 0 in the default map.

Reply:
It's probably a good idea to do this. However, there are a few things that make the real version more complex:

* The 48kb SRAM is actually split between 40kb SRAM at `0x2000 0000 - 0x2000 9FFF` and 8k CCRAM at `0x1000 0000 - 0x1000 1FFF`. For multiple reasons (including my inadequate knowledge in what the heck CCRAM is used for), I decided to try to simplify.
* I don't actually own one of these, and was confused when I saw this in that the Flash memory is listed twice in the reference manual memory map :)

<img width="843" alt="image" src="https://user-images.githubusercontent.com/609906/83319410-39211c80-a20c-11ea-9fb8-b1fc8445aa40.png">